runtime/interpreter/interpreter.go

```diff
@@ -1905,10 +1992,119 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
 			return ConvertAddress(interpreter, value, locationRange)
 		}
 
+	case *sema.ConstantSizedType, *sema.VariableSizedType:
+		if arrayValue, isArray := value.(*ArrayValue); isArray && !valueType.Equal(unwrappedTargetType) {
+
+			arrayStaticType := interpreter.convertStaticType(arrayValue.StaticType(interpreter), unwrappedTargetType).(ArrayStaticType)
+			targetElementType := interpreter.MustConvertStaticToSemaType(arrayStaticType.ElementType())
+
+			array := arrayValue.array
+
+			iterator, err := array.Iterator()
+			if err != nil {
+				panic(errors.NewExternalError(err))
+			}
+
+			return NewArrayValueWithIterator(
+				interpreter,
+				arrayStaticType,
+				arrayValue.GetOwner(),
+				array.Count(),
+				func() Value {
+					element, err := iterator.Next()
+					if err != nil {
+						panic(errors.NewExternalError(err))
+					}
+					if element == nil {
+						return nil
+					}
+
+					value := MustConvertStoredValue(interpreter, element)
+					valueType := interpreter.MustConvertStaticToSemaType(value.StaticType(interpreter))
+					return interpreter.convert(value, valueType, targetElementType, locationRange)
+				},
+			)
+		}
+
+	case *sema.DictionaryType:
```

Comment:
So for arrays and dictionaries, this function didn't create new value instances previously, but now it does.

Given this can be expensive, is it possible to ensure that this conversion happens only for reference types (containers with references inside), but not for other types?

Reply:
Also, have we considered the alternative solution: making arrays and dictionaries non-covariant?
Might need a FLIP though

Reply:
I think we should be able to do this only when the converted static type doesn't match the original, as that would indicate a reference type within has changed. Implemented this in 9bae858d79ee52359fd88c7ce037584f78f532bd